File: java/tst/TestBlob.java
import org.garret.perst.*;

import java.io.*;

public class TestBlob { 
    public static void main(String[] args) throws Exception { 
        Storage db = StorageFactory.getInstance().createStorage();
        db.open("testblob.dbs");
        Index root = (Index)db.getRoot();
        byte[] buf = new byte[1024];
        int rc;
        File dir = new File(".");
        String[] files = dir.list();
        if (root == null) { 
            root = db.createIndex(String.class, true);
            db.setRoot(root);
            for (int i = 0; i < files.length; i++) { 
                if (files[i].endsWith(".java")) {
                    FileInputStream in = new FileInputStream(files[i]);
                    Blob blob = db.createBlob();                    
                    OutputStream out = blob.getOutputStream(false);
                    while ((rc = in.read(buf)) > 0) { 
                        out.write(buf, 0, rc);
                    }
                    root.put(files[i], blob);   
                    in.close();
                    out.close();
                }
            }
            System.out.println("Database is initialized");
        } 
        for (int i = 0; i < files.length; i++) { 
            if (files[i].endsWith(".java")) {
                byte[] buf2 = new byte[1024];
                Blob blob = (Blob)root.get(files[i]);
                if (blob == null) {
                    System.err.println("File " + files[i] + " not found in database");
                    continue;
                }
                InputStream bin = blob.getInputStream();
                InputStream fin = new FileInputStream(files[i]);
                while ((rc = fin.read(buf)) > 0) { 
                    int rc2 = bin.read(buf2);
                    if (rc != rc2) {
                        System.err.println("Different file size: " + rc + " .vs. " + rc2);
                        break;
                    }
                    while (--rc >= 0 && buf[rc] == buf2[rc]);
                    if (rc >= 0) { 
                        System.err.println("Content of the files is different: " + buf[rc] + " .vs. " + buf2[rc2]);
                        break;
                    }
                }
                bin.close();
                fin.close();
            }            
        }
        System.out.println("Verification completed");
        db.close();
    }
}

File: csharp/TestBlob/TestBlob.cs
using System;
using Perst;
using System.IO;

public class TestBlob 
{ 
    public static void Main(string[] args) 
    { 
        Storage db = StorageFactory.Instance.CreateStorage();
        db.Open("testblob.dbs");
        byte[] buf = new byte[1024];
        int rc;
        string[] files = Directory.GetFiles("\\Perst.NET\\src\\impl", "*.cs");
        Index root = (Index)db.Root;
        if (root == null) 
        { 
            root = db.CreateIndex(typeof(string), true);
            db.Root = root;
            foreach (String file in files) 
            { 
                FileStream fin = new FileStream(file, FileMode.Open, FileAccess.Read);
                Blob blob = db.CreateBlob();                    
                Stream bout = blob.GetStream();
                while ((rc = fin.Read(buf, 0, buf.Length)) > 0) 
                { 
                    bout.Write(buf, 0, rc);
                }
                root[file] = blob; 
                fin.Close();
                bout.Close();   
            }
            Console.WriteLine("Database is initialized");
        } 
        foreach (String file in files) 
        {
            byte[] buf2 = new byte[1024];
            Blob blob = (Blob)root[file];
            if (blob == null) 
            {
                Console.WriteLine("File " + file + " not found in database");
                continue;
            }
            Stream bin = blob.GetStream();
            FileStream fin = new FileStream(file, FileMode.Open, FileAccess.Read);
            while ((rc = fin.Read(buf, 0, buf.Length)) > 0) 
            { 
                int rc2 = bin.Read(buf2, 0, buf2.Length);
                if (rc != rc2) 
                {
                    Console.WriteLine("Different file size: " + rc + " .vs. " + rc2);
                    break;
                }
                while (--rc >= 0 && buf[rc] == buf2[rc]);
                if (rc >= 0) 
                { 
                    Console.WriteLine("Content of the files is different: " + buf[rc] + " .vs. " + buf2[rc]);
                    break;
                }
            }
            fin.Close();
            bin.Close();
        }            
        Console.WriteLine("Verification completed");
        db.Close();
    }
}

